README.md
midifilter.lv2
==============

LV2 plugins to filter MIDI events.

So far 27 MIDI event filters have been implemented:

*   CC2Note -- translate control-commands to note-on/offmessages
*   Channel Filter -- discard messages per Channel
*   Channel Map -- map any MIDI-channel to another MIDI-channel
*   Enforce Scale -- force midi notes on given musical scale
*   Eventblocker -- notch style message filter. Suppress specific messages
*   Keyrange -- discard notes-on/off events outside an give range
*   Keysplit -- change midi-channel number depending on note (and optionally transpose)
*   MapCC -- change one control message into another
*   Mapscale -- flexible 12-tone map
*   Chord -- harmonizer - create chords from a single note in a given musical scale
*   Delay -- delay MIDI events with optional radomization
*   Dup -- unisono - duplicate MIDI events from one channel to another
*   Strum -- arpeggio effect intended to simulate strumming a stringed instrument (e.g. guitar)
*   Transpose -- chromatic transpose MIDI notes
*   Legato -- hold a note until the next note arrives
*   NoSensing -- strip MIDI Active-Sensing events
*   NoDup -- MIDI duplicate blocker. Filter out overlapping note on/off and duplicate messages
*   Note2CC -- convert MIDI note-on messages to control change messages
*   NoteToggle -- toggle notes: play a note to turn it on, play it again to turn it off
*   nTabDelay -- repeat notes N times (incl tempo-ramps -- eurotechno hell yeah)
*   Passthru -- no operation, just pass the MIDI event through (example plugin)
*   Quantize -- live midi event quantization
*   Velocity Randomizer -- randomly change velocity of note-on events
*   ScaleCC -- modify the value (data-byte) of a MIDI control change message
*   Sostenuto -- delay note-off messages, emulate a piano sostenuto pedal
*   Velocity Range -- filter MIDI note events according to velocity
*   Velocity Scale -- modify note velocity by constant factor and offset

Install
-------

Compiling the plugins requires LV2 SDK, gnu-make and a c-compiler.

```bash
  git clone git://github.com/x42/midifilter.lv2.git
  cd midifilter.lv2
  make
  sudo make install PREFIX=/usr
```


ChangeLog
2015-06-30 (0.4.2)  Robin Gareus <robin@gareus.org>
* honor playback speed in BBT calc & fix zero speed
* add LV2 version
* fix ranges and explicitly use int

2015-06-08 (0.4.1)  Robin Gareus <robin@gareus.org>
* key map/transpoe: use an enum with unit semitone

2015-05-23 (0.4.0)  Robin Gareus <robin@gareus.org>
* portablility updates (OSX, windows)
* fix potential array overflow in delaylines

2014-01-06 (0.3.0)  Robin Gareus <robin@gareus.org>
* add CC-map filter
* add CC-to-note
* sostenuto: allow to use CC64 to control pedal-state

2013-11-19 (0.2.0)  Robin Gareus <robin@gareus.org>
* new: midiCC value scale/offset/invert filter
* new: "Note to CC" plugin
* add inversion point feature to miditranspose plugin
* fix various warning with RDF (missing semicolon, ranges)
* add self-doc script using lv2toweb -> https://github.com/x42/midifilter.lv2

2013-10-21 (0.1.3)  Robin Gareus <robin@gareus.org>
* remove multi-line comment (cpp warning
* updated README
* fix typo: rename ntabdelay -> ntapdelay

2013-09-15 (0.1.2)  Robin Gareus <robin@gareus.org>
* add plugin description/doc
* update FSF address in GPL header

2013-06-16 (0.1.1)  Robin Gareus <robin@gareus.org>
* add some more built-in doc

2013-06-09 (0.1.0)  Robin Gareus <robin@gareus.org>
* fix transpose plugin -- check for parameter change in every process cycle
* add 'enforce scale' plugin
* many more new plugins.

2013-05-24 (0.0.3)  Robin Gareus <robin@gareus.org>
* new plugins:
- delayline w/ optional random delay
- velocity randomizer
- normrandvel - velocity filter with normal distribution
- sostenuto - piano pedal effect
- channel filter

2013-05-23 (0.0.2)  Robin Gareus <robin@gareus.org>
* first plugins: channel-map, velocity, duplicate/unison
* transpose plugin
* add keysplit w/transpose plugin
* filter active-sensing

2013-05-20 (0.0.1)  Robin Gareus <robin@gareus.org>
* initial framework, passthru example


ttf.h
#ifndef _TTF_H_
#define _TTF_H_

/* cfg-port offsets 0,1, are midi ports, . */
#define ADDP_0 3
#define ADDP_1 4
#define ADDP_2 5
#define ADDP_3 6
#define ADDP_4 7
#define ADDP_5 8
#define ADDP_6 9
#define ADDP_7 10
#define ADDP_8 11
#define ADDP_9 12
#define ADDP_10 13
#define ADDP_11 14
#define ADDP_12 15
#define ADDP_13 16
#define ADDP_14 17
#define ADDP_15 18

#define PORTIDX(x) ADDP_##x

#define MAINTAINER <HTTPP/gareus.org/rgareus#me>
#define MIDIEXTURI <HTTPP/lv2plug.in/ns/ext/midi#MidiEvent>

#define TTF_DEF(DOAPNAME, INSUPPORTS) \
	a lv2:Plugin, lv2:FilterPlugin ; \
	doap:name DOAPNAME ; \
	@VERSION@ \
	doap:license <HTTPP/usefulinc.com/doap/licenses/gpl> ; \
  lv2:project <HTTPP/gareus.org/oss/lv2/midifilter> ; \
	lv2:optionalFeature lv2:hardRTCapable ; \
	lv2:requiredFeature urid:map ; \
	lv2:port \
	[ \
		a atom:AtomPort , \
			lv2:InputPort ; \
		atom:bufferType atom:Sequence ; \
		atom:supports MIDIEXTURI INSUPPORTS; \
		lv2:index 0 ; \
		lv2:symbol "midiin" ; \
		lv2:name "MIDI In" \
	] , [ \
		a atom:AtomPort , \
			lv2:OutputPort ; \
		atom:bufferType atom:Sequence ; \
		atom:supports MIDIEXTURI ; \
		lv2:index 1 ; \
		lv2:symbol "midiout" ; \
		lv2:name "MIDI Out"; \
	] , [ \
		a lv2:OutputPort, \
			lv2:ControlPort ; \
		lv2:name "latency" ; \
		lv2:index 2 ; \
		lv2:symbol "latency" ; \
		lv2:minimum 0 ; \
		lv2:maximum 192000 ; \
		lv2:portProperty lv2:reportsLatency, lv2:integer, pprops:notOnGUI; \
	]

#define TTF_DEFAULTDEF(DOAPNAME) TTF_DEF(DOAPNAME,)

#define TTF_PORT(TYPE, IDX, SYM, DESC, VMIN, VMAX, VDFLT, ATTR) \
	[ \
    a TYPE, \
      lv2:ControlPort ; \
    lv2:index PORTIDX(IDX) ; \
    lv2:symbol SYM ; \
    lv2:name DESC; \
    lv2:minimum VMIN ; \
    lv2:maximum VMAX ; \
    lv2:default VDFLT; \
    ATTR \
  ]

#define TTF_IPORT(IDX, SYM, DESC, VMIN, VMAX, VDFLT, ATTR) \
	TTF_PORT(lv2:InputPort, IDX, SYM, DESC, VMIN, VMAX, VDFLT, ATTR)

#define TTF_IPORTFLOAT(IDX, SYM, DESC, VMIN, VMAX, VDFLT) \
	TTF_IPORT(IDX, SYM, DESC, VMIN, VMAX, VDFLT, )

#define TTF_IPORTINT(IDX, SYM, DESC, VMIN, VMAX, VDFLT) \
	TTF_IPORT(IDX, SYM, DESC, VMIN, VMAX, VDFLT, lv2:portProperty lv2:integer)

#define TTF_IPORTTOGGLE(IDX, SYM, DESC, VDFLT) \
	TTF_IPORT(IDX, SYM, DESC, 0, 1, VDFLT, lv2:portProperty lv2:integer; lv2:portProperty lv2:toggled)

#define PORTENUM16 \
	lv2:scalePoint [ rdfs:label "01" ; rdf:value  1 ] ; \
	lv2:scalePoint [ rdfs:label "02" ; rdf:value  2 ] ; \
	lv2:scalePoint [ rdfs:label "03" ; rdf:value  3 ] ; \
	lv2:scalePoint [ rdfs:label "04" ; rdf:value  4 ] ; \
	lv2:scalePoint [ rdfs:label "05" ; rdf:value  5 ] ; \
	lv2:scalePoint [ rdfs:label "06" ; rdf:value  6 ] ; \
	lv2:scalePoint [ rdfs:label "07" ; rdf:value  7 ] ; \
	lv2:scalePoint [ rdfs:label "08" ; rdf:value  8 ] ; \
	lv2:scalePoint [ rdfs:label "09" ; rdf:value  9 ] ; \
	lv2:scalePoint [ rdfs:label "10" ; rdf:value 10 ] ; \
	lv2:scalePoint [ rdfs:label "11" ; rdf:value 11 ] ; \
	lv2:scalePoint [ rdfs:label "12" ; rdf:value 12 ] ; \
	lv2:scalePoint [ rdfs:label "13" ; rdf:value 13 ] ; \
	lv2:scalePoint [ rdfs:label "14" ; rdf:value 14 ] ; \
	lv2:scalePoint [ rdfs:label "15" ; rdf:value 15 ] ; \
	lv2:scalePoint [ rdfs:label "16" ; rdf:value 16 ] ; \
	lv2:portProperty lv2:integer; lv2:portProperty lv2:enumeration;

#define SPX(LBL,VAL) \
	lv2:scalePoint [ rdfs:label LBL ; rdf:value VAL ] ;

#define PORTENUMZ(ZEROLABEL) \
	lv2:scalePoint [ rdfs:label ZEROLABEL; rdf:value  0 ] ; \
	PORTENUM16

#define DOC_CHANZ \
	rdfs:comment "MIDI Channel number 1..16 (0: disable channel)"

#define DOC_CHANF \
	rdfs:comment "MIDI Channel (1..16) on which the filter is active; 0: any). Data on other channels is passed though unmodified."

#define NOTENAMES \
	lv2:scalePoint [ rdfs:label "C"  ; rdf:value 0 ] ; \
	lv2:scalePoint [ rdfs:label "C#" ; rdf:value 1 ] ; \
	lv2:scalePoint [ rdfs:label "D"  ; rdf:value 2 ] ; \
	lv2:scalePoint [ rdfs:label "D#" ; rdf:value 3 ] ; \
	lv2:scalePoint [ rdfs:label "E"  ; rdf:value 4 ] ; \
	lv2:scalePoint [ rdfs:label "F"  ; rdf:value 5 ] ; \
	lv2:scalePoint [ rdfs:label "F#" ; rdf:value 6 ] ; \
	lv2:scalePoint [ rdfs:label "G"  ; rdf:value 7 ] ; \
	lv2:scalePoint [ rdfs:label "G#" ; rdf:value 8 ] ; \
	lv2:scalePoint [ rdfs:label "A"  ; rdf:value 9 ] ; \
	lv2:scalePoint [ rdfs:label "A#" ; rdf:value 10 ] ; \
	lv2:scalePoint [ rdfs:label "B"  ; rdf:value 11 ] ; \
	lv2:portProperty lv2:integer; lv2:portProperty lv2:enumeration;

#define NOTENAMESOFF \
	lv2:scalePoint [ rdfs:label "OFF" ; rdf:value -1 ] ; \
	NOTENAMES

#endif

/* variable part */

#ifdef MFD_FLT
#undef MFD_FLT
#endif

#ifdef MX_FILTER
#define MFD_FLT(ID, FNX) \
	else if (!strcmp(descriptor->URI, MFP_URI "#" # FNX)) { self->filter_fn = filter_midi_ ## FNX; filter_init_ ## FNX(self); }

#elif (defined MX_DESC)

#define MFD_FLT(ID, FNX) \
	MF_DESCRIPTOR(ID, "" # FNX)

#elif (defined MX_MANIFEST)

#define MFD_FLT(ID, FNX) \
  <HTTPP/gareus.org/oss/lv2/midifilterHASH ## FNX> \
	a lv2:Plugin ; \
	lv2:binary <@LV2NAME@@LIB_EXT@>  ; \
	rdfs:seeAlso <@LV2NAME@.ttl> . \

#else

#define MFD_FLT(ID, FNX)

#endif
